elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts

```diff
@@ -239,9 +260,171 @@ export class SparseMerkleTree {
+				binaryBitmap: bufferToBinaryString(EMPTY_HASH),
+				siblingHashes: [],
+				ancestorHashes: [],
+			};
+		}
+
+		let h = 0;
+		const siblingHashes = [];
+		const ancestorHashes = [];
+		let binaryBitmap = '';
+		const binaryKey = binaryExpansion(queryKey, this.keyLength);
+
+		while (currentNode instanceof Branch) {
+			ancestorHashes.push(currentNode.hash);
+			const d = binaryKey.charAt(h);
+			let currentNodeSibling: TreeNode = new Empty();
+			if (d === '0') {
+				currentNodeSibling = await this.getNode(currentNode.rightHash);
+				currentNode = await this.getNode(currentNode.leftHash);
+			} else if (d === '1') {
+				currentNodeSibling = await this.getNode(currentNode.leftHash);
+				currentNode = await this.getNode(currentNode.rightHash);
+			}
+
+			if (currentNodeSibling instanceof Empty) {
+				binaryBitmap = `0${binaryBitmap}`;
+			} else {
+				binaryBitmap = `1${binaryBitmap}`;
+				siblingHashes.push(currentNodeSibling.hash);
+			}
+			h += 1;
+		}
+
+		if (currentNode instanceof Empty) {
+			// exclusion proof
+			return {
+				siblingHashes,
+				ancestorHashes,
+				binaryBitmap,
+				key: queryKey,
+				value: EMPTY_VALUE,
+			};
+		}
+
+		if (
+			currentNode instanceof Leaf &&
+			currentNode.key.toString('hex') !== queryKey.toString('hex')
+		) {
+			// exclusion proof
+			ancestorHashes.push(currentNode.hash); // in case the leaf is sibling to another node
+			return {
+				siblingHashes,
+				ancestorHashes,
+				binaryBitmap,
+				key: currentNode.key,
+				value: currentNode.value,
+			};
+		}
+		if (
+			currentNode instanceof Leaf &&
+			currentNode.key.toString('hex') === queryKey.toString('hex')
+		) {
+			// inclusion proof
+			ancestorHashes.push(currentNode.hash); // in case the leaf is sibling to another node
+			return {
+				siblingHashes,
+				ancestorHashes,
+				binaryBitmap,
+				key: currentNode.key,
+				value: currentNode.value,
+			};
+		}
+		return {
+			key: queryKey,
+			value: EMPTY_VALUE,
+			binaryBitmap: bufferToBinaryString(EMPTY_HASH),
+			siblingHashes: [],
+			ancestorHashes: [],
+		};
+	};
+
+	public generateMultiProof = async (queryKeys: Buffer[]): Promise<Proof> => {
+		const partialQueries: SingleProof[] = [];
+		for (const queryKey of queryKeys) {
+			const query = await this.generateSingleProof(queryKey);
+			partialQueries.push(query);
+		}
+
+		const queries: Query[] = [...partialQueries].map(sp => ({
+			bitmap: binaryStringToBuffer(sp.binaryBitmap),
+			key: sp.key,
+			value: sp.value,
+		}));
+		const siblingHashes: Buffer[] = [];
+		const ancestorHashes = [...partialQueries].map(sp => sp.ancestorHashes).flat();
+		let sortedQueries: QueryWithHeight[] = [...partialQueries].map(sp => ({
+			binaryBitmap: sp.binaryBitmap,
+			key: sp.key,
+			value: sp.value,
+			siblingHashes: sp.siblingHashes,
+			height: sp.binaryBitmap.length,
+		}));
+		sortedQueries = sortByBitmapAndKey(sortedQueries);
+
+		while (sortedQueries.length > 0) {
+			const sp = sortedQueries.shift()!;
```

Comment:
`shift` is very slow, so prefer not to use them. it re-indexes the whole array

Reply:
Any suggestions ?

Reply:
Maybe keep track of the index of `sortedQuery` might be better?
but maybe not so important if the sortedQueries doesn't grow much in size.

Reply:
https://github.com/LiskHQ/lisk-sdk/issues/6597